### PR TITLE
Fix integral AABB issue

### PR DESCRIPTION
**Pull request by oktomus**
I haven't set `extent` and `volume` to virtual so that `static invalid` could be used. But this can be changed if we need.

### CODE REVIEW

File: src/appleseed/foundation/math/aabb.h [hunk 1]
```diff
@@ -191,31 +164,101 @@ template <typename T, size_t N> AABB<T, N>& operator*=(AABB<T, N>& lhs, const T
 // Compute the surface area of a 3D bounding box.
 template <typename T> T half_surface_area(const AABB<T, 3>& bbox);
 template <typename T> T surface_area(const AABB<T, 3>& bbox);
+//
```
Comment (dictoon) [inline src/appleseed/foundation/math/aabb.h]: Add two blank lines before block comment.

File: src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp [hunk 2]
```diff
@@ -557,7 +557,7 @@ namespace
               , m_timer_start_value(m_timer.read())
             {
                 const Vector2u crop_window_extent = m_project.get_frame()->get_crop_window().extent();
-                const size_t pixel_count = (crop_window_extent.x + 1) * (crop_window_extent.y + 1);
+                const size_t pixel_count = (crop_window_extent.x) * (crop_window_extent.y);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp]: Remove useless parentheses.

File: src/appleseed/foundation/math/aabb.h [hunk 3]
```diff
@@ -191,31 +164,101 @@ template <typename T, size_t N> AABB<T, N>& operator*=(AABB<T, N>& lhs, const T
...
+// Floating N-dimensional axis-aligned bounding box [min, max] class.
+//
+
+template <typename T, size_t N>
+class FloatingAABB
+  : public AABB<T, N>
+{
+  public:
+    using AABB<T, N>::AABB;
```
Comment (dictoon) [inline src/appleseed/foundation/math/aabb.h]: What's this?

Comment (oktomus) [reply src/appleseed/foundation/math/aabb.h]: To inherit constructors

File: src/appleseed/foundation/math/intersection/rayaabb.h [hunk 4]
```diff
@@ -140,9 +140,9 @@ inline bool intersect(
 
 template <>
 inline bool intersect<float>(
-    const Ray3f&            ray,
-    const RayInfo3f&        ray_info,
-    const AABB3f&           bbox)
+    const Ray3f&        ray,
```
Comment (dictoon) [inline src/appleseed/foundation/math/intersection/rayaabb.h]: Vertically align arguments with neighbor functions..

File: src/appleseed/foundation/math/intersection/rayaabb.h [hunk 5]
```diff
@@ -177,9 +177,9 @@ inline bool intersect<float>(
 
 template <>
 inline bool intersect<double>(
-    const Ray3d&            ray,
-    const RayInfo3d&        ray_info,
-    const AABB3d&           bbox)
+    const Ray3d&        ray,
```
Comment (dictoon) [inline src/appleseed/foundation/math/intersection/rayaabb.h]: Vertically align arguments with neighbor functions..

File: src/appleseed/foundation/math/aabb.h (hunk 3 above)
Comment (dictoon) [reply src/appleseed/foundation/math/aabb.h]: Unlikely to work on VS 2012. I'll test.

Comment (oktomus) [reply src/appleseed/foundation/math/aabb.h]: Thanks

Comment (dictoon) [reply src/appleseed/foundation/math/aabb.h]: Seems like it's supported, I didn't know, cool!

File: src/appleseed/foundation/math/aabb.h [hunk 6]
```diff
@@ -465,75 +423,118 @@ inline size_t AABB<T, N>::rank() const
...
+    lhs.min *= rhs;
+    lhs.max *= rhs;
+    return lhs;
+}
 
-    return T(0.5) * diameter();
+template <typename T, size_t N>
+inline Vector<T, N> AABBBase<T, N>::extent() const
+{
+    assert(this->is_valid());
```
Comment (dictoon) [inline src/appleseed/foundation/math/aabb.h]: Are `this->` really necessary? If not, please remove them (everywhere).

File: src/appleseed/foundation/math/aabb.h [hunk 7]
```diff
@@ -174,19 +147,69 @@ class AABB
...
-template <typename T, size_t N> AABB<T, N>& operator*=(AABB<T, N>& lhs, const T rhs);
+template <typename T, size_t N> AABBBase<T, N>  operator+ (const AABBBase<T, N>& lhs, const AABBBase<T, N>& rhs);
+template <typename T, size_t N> AABBBase<T, N>  operator* (const AABBBase<T, N>& lhs, const T rhs);
+template <typename T, size_t N> AABBBase<T, N>  operator* (const T lhs, const AABBBase<T, N>& rhs);
+template <typename T, size_t N> AABBBase<T, N>& operator+=(AABBBase<T, N>& lhs, const AABBBase<T, N>& rhs);
+template <typename T, size_t N> AABBBase<T, N>& operator*=(AABBBase<T, N>& lhs, const T rhs);
+
+
+//
+// N-dimensional floating point axis-aligned bounding box [min, max] class.
```
Comment (dictoon) [inline src/appleseed/foundation/math/aabb.h]: *floating-point*

File: src/appleseed/foundation/math/aabb.h [hunk 8]
```diff
@@ -174,19 +147,69 @@ class AABB
...
+    ValueType radius() const;
+
+    // Return the square radius of the bounding box.
+    ValueType square_radius() const;
+
+    // Compute the center of the bounding box.
+    VectorType center() const;
+    ValueType center(const size_t dim) const;
+
+    VectorType extent() const;
```
Comment (dictoon) [inline src/appleseed/foundation/math/aabb.h]: Copy comments from base class for these methods.

File: src/appleseed/foundation/meta/tests/test_aabb.cpp [hunk 9]
```diff
@@ -387,6 +426,21 @@ TEST_SUITE(Foundation_Math_AABB)
         EXPECT_FEQ(Vector3d(5.0, 7.0, 9.0), bbox.extent());
     }
 
+    TEST_CASE(TestExtentOnIntegralAABB)
+    {
+        const AABB3i bbox(
+            Vector3i(-1, -2, -3),
+            Vector3i(4, 5, 6));
+
+        EXPECT_FEQ(Vector3i(6, 8, 10), bbox.extent());
```
Comment (dictoon) [inline src/appleseed/foundation/meta/tests/test_aabb.cpp]: Use `EXPECT_EQ()` since you're working with integral types (same change in other tests).

File: src/appleseed/foundation/meta/tests/test_aabb.cpp [hunk 10]
```diff
@@ -387,6 +426,21 @@ TEST_SUITE(Foundation_Math_AABB)
...
+
+        const AABB3i bbox2(
+            Vector3i(0, -1, 1),
+            Vector3i(4, 4, 4));
+
+        EXPECT_FEQ(Vector3i(5, 6, 4), bbox2.extent());
```
Comment (dictoon) [inline src/appleseed/foundation/meta/tests/test_aabb.cpp]: One test is enough, you can remove this one unless it's testing another corner case in which case it's best to make it a separate test with an explicit name.

File: src/appleseed/foundation/meta/tests/test_aabb.cpp [hunk 11]
```diff
@@ -536,4 +599,15 @@ TEST_SUITE(Foundation_Math_AABB)
         EXPECT_FALSE(bbox1 != bbox2);
         EXPECT_TRUE(bbox1 != bbox3);
     }
+
+    TEST_CASE(TestOneByOneOnIntegralAABB)
```
Comment (dictoon) [inline src/appleseed/foundation/meta/tests/test_aabb.cpp]: Maybe `TestExtentOn1x1x1IntegralAABB`? Probably best moved next to the other tests for `extent()`.

File: src/appleseed/foundation/math/aabb.h (hunk 6 above)
Comment (oktomus) [reply src/appleseed/foundation/math/aabb.h]: Yes they are, doesn't compile otherwise
